poly1305/src/avx2.rs

```diff
@@ -0,0 +1,157 @@
+//! AVX2 implementation of the Poly1305 state machine.
+
+// The State struct and its logic was originally derived from Goll and Gueron's AVX2 C
+// code:
+//     [Vectorization of Poly1305 message authentication code](https://ieeexplore.ieee.org/document/7113463)
+//
+// which was sourced from Bhattacharyya and Sarkar's modified variant:
+//     [Improved SIMD Implementation of Poly1305](https://eprint.iacr.org/2019/842)
+//     https://github.com/Sreyosi/Improved-SIMD-Implementation-of-Poly1305
+//
+// The logic has been extensively rewritten and documented, and several bugs in the
+// original C code were fixed.
+//
+// Note that State only implements the original Goll-Gueron algorithm, not the
+// optimisations provided by Bhattacharyya and Sarkar. The latter require the message
+// length to be known, which is incompatible with the streaming API of UniversalHash.
+
+use universal_hash::generic_array::GenericArray;
+
+use crate::{Block, Key, Tag, BLOCK_SIZE};
+
+mod helpers;
+use self::helpers::*;
+
+const BLOCK_X4_SIZE: usize = BLOCK_SIZE * 4;
+
+#[derive(Clone)]
+struct Initialized {
+    p: Aligned4x130,
+    m: SpacedMultiplier4x130,
+    r4: PrecomputedMultiplier,
+}
+
+#[derive(Clone)]
+pub(crate) struct State {
+    k: AdditionKey,
+    r1: PrecomputedMultiplier,
+    r2: PrecomputedMultiplier,
+    initialized: Option<Initialized>,
+    cached_blocks: [u8; BLOCK_X4_SIZE],
```

Comment:
A bit more of a meta-note, but I'm wondering if the `UniversalHash` trait should have something more like `ParBlocks` on `UniversalHash` which can be used to signal to the consumer of the trait how many blocks the underlying implementation can work on at a time:

https://docs.rs/block-cipher/0.8.0/block_cipher/trait.BlockCipher.html#associatedtype.ParBlocks

Reply:
Opened an issue about this for the upstream `universal-hash` crate: https://github.com/RustCrypto/traits/issues/289

Let's continue discussion there. Otherwise I'd say this is fine for now and we should figure out a general solution upstream.